File: src/main/java/unidue/rc/ui/mixins/Confirm.java
package unidue.rc.ui.mixins;

/*
 * #%L
 * Semesterapparate
 * $Id:$
 * $HeadURL:$
 * %%
 * Copyright (C) 2014 - 2015 Universitaet Duisburg Essen
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import org.apache.tapestry5.BindingConstants;
import org.apache.tapestry5.ClientElement;
import org.apache.tapestry5.ComponentResources;
import org.apache.tapestry5.annotations.AfterRender;
import org.apache.tapestry5.annotations.Import;
import org.apache.tapestry5.annotations.InjectContainer;
import org.apache.tapestry5.annotations.Parameter;
import org.apache.tapestry5.ioc.Messages;
import org.apache.tapestry5.ioc.annotations.Inject;
import org.apache.tapestry5.json.JSONObject;
import org.apache.tapestry5.services.javascript.JavaScriptSupport;

/**
 * A mixin used to attach a JavaScript confirmation box to the onclick
 * event of any component that implements ClientElement.
 * Created by nils on 09.06.15.
 */
@Import(library = { "context:js/mixins/confirm.js" })
public class Confirm {
    /**
     * Confirmation message to display.
     */
    @Parameter(defaultPrefix = BindingConstants.LITERAL)
    private String message;

    /**
     * Dialog box title.
     */
    @Parameter(defaultPrefix = BindingConstants.LITERAL)
    private String title;

    /**
     * Validation Button text.
     */
    @Parameter(defaultPrefix = BindingConstants.LITERAL)
    private String validationMsg;

    /**
     * Cancel Button text.
     */
    @Parameter(defaultPrefix = BindingConstants.LITERAL)
    private String cancelMsg;

    /**
     * Confirmation box height.
     */
    @Parameter(value = "240", defaultPrefix = BindingConstants.LITERAL)
    private int height;

    /**
     * Confirmation box width.
     */
    @Parameter(value = "320", defaultPrefix = BindingConstants.LITERAL)
    private int width;
    /**
     * If this parameter is set to <i>true</i>, the user can't interact with the application while the
     * confirmation box is displayed.
     */
    @Parameter(value = "true", defaultPrefix = BindingConstants.LITERAL)
    private boolean isModal;

    /**
     * If this parameter is set to <i>true</i>, the user can dynamically resize the confirmation box.
     */
    @Parameter(value = "false", defaultPrefix = BindingConstants.LITERAL)
    private boolean isResizable;

    /**
     * If this parameter is set to <i>true</i>, the user can drag the confirmation box.
     */
    @Parameter(value = "true", defaultPrefix = BindingConstants.LITERAL)
    private boolean isDraggable;

    /**
     * If this parameter is set to <i>true</i>, default javascript alert is used. Otherwise, jquery dialog box
     * is used, and can be customized with several parameters.
     * <p>
     * Usable parameters in both configurations :
     * message
     * <p>
     * Usable parameters when useDefaultConfirm = false:
     * title, validationMsg, cancelMsg, isModal, isResizable, isDraggable
     * <p>
     * validationMsg, cancelMsg, isModal, isResizable, height
     */
    @Parameter(value = "false", defaultPrefix = BindingConstants.LITERAL)
    private boolean useDefaultConfirm;


    /**
     * since 3.4.3
     * The Confirm parameters you want to override.
     * <p>
     * This will be used only if seDefaultConfirm = false (defaultValue):
     */
    @Parameter
    private JSONObject params;

    //Injected services.

    @Inject
    private JavaScriptSupport javaScriptSupport;

    @InjectContainer
    private ClientElement element;

    @Inject
    private ComponentResources resources;

    @Inject
    private Messages messageProvider;

    @AfterRender
    public void afterRender() {
        JSONObject config = new JSONObject();

        String clientId = element.getClientId();

        config.put("id", clientId);

        config.put("title", title != null ? title : messageProvider.get("default-confirm-title"));
        config.put("message", message != null ? message : messageProvider.get("default-confirm-message"));
        config.put("validationMsg", validationMsg != null ? validationMsg : messageProvider.get("default-valid-label"));
        config.put("cancelMsg", cancelMsg != null ? cancelMsg : messageProvider.get("default-cancel-label"));

        config.put("useDefaultConfirm", useDefaultConfirm);
        config.put("isModal", isModal);
        config.put("isResizable", isResizable);
        config.put("isDraggable", isDraggable);
        config.put("height", height);
        config.put("width", width);


        /*
         * We will merge the default JSON Object with the params parameter
         */
        if (resources.isBound("params")) {
            merge(config, params);
        }

        javaScriptSupport.addInitializerCall("confirm", config);
    }

    static JSONObject merge(JSONObject obj1, JSONObject obj2) {
        if (obj1 == null)
            return null;

        if (obj2 == null)
            return obj1;

        for (String key : obj2.keys()) {
            obj1.put(key, obj2.get(key));
        }

        return null;
    }
}

File: src/main/webapp/js/mixins/confirm.js
/*
 * Copyright (C) 2014 - 2016 Universitaet Duisburg-Essen (semapp|uni-due.de)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Created by nils on 09.06.15.
 */
(function($) {

    T5.extendInitializers(function() {

        function init(specs) {

            //JQuery dialog box configuration, if used.
            if (!specs.useDefaultConfirm) {
                var dialogBox = $('<div id=\'dialogConfirmationJQuery\' />').html(specs.message).dialog({
                    autoOpen : false,
                    resizable : specs.isResizable,
                    height : specs.height,
                    resize : 'auto',
                    title : specs.title,
                    modal : specs.isModal,
                    draggable : specs.isDraggable,
                    buttons : [ {
                        text : specs.validationMsg,
                        click : function() {
                            $(this).dialog("close");

                            trigger($("#" + specs.id));
                        }
                    }, {
                        text : specs.cancelMsg,
                        click : function() {
                            $(this).dialog("close");
                        }
                    } ]
                });
            }

            $("#" + specs.id).click(function(event) {
                if (specs.useDefaultConfir) {
                    //Default javascript confirmation box.
                    return confirm(specs.message);
                } else {
                    //JQuery confirmation box : we stop the event propagation before displaying the box.
                    //Otherwise the redirection will occur while confirmation box is displayed.

                    event.preventDefault();
                    event.stopImmediatePropagation();

                    dialogBox.dialog("open");
                }
            });
        }

        /**
         * Trigger url redirection or firm submission from element.
         *
         * @param HTML element
         */
        function trigger(element) {
            var tagName = element.prop("tagName");
            switch (tagName) {
                //Simple link (pagelink, actionlink, etc...)
                case "A":
                    var href = element.prop("href");
                    if (href != undefined) {
                        var urlSuffix = href.substring(href.lastIndexOf('.') + 1);
                        if (urlSuffix == element.prop("id")) {
                            //ActionLink
                            element.trigger(Tapestry.TRIGGER_ZONE_UPDATE_EVENT);
                        } else {
                            window.location.href = href;
                        }
                    }
                    break;
                //submit button.
                case "INPUT":
                    element.parents("form").submit();
                    break;
                default:
                    break;
            }
        }

        return {
            confirm : init
        }
    });
})(jQuery);